keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,746 @@
+	SetMachineState(memoryMap NodeMap)
+	// SetMemoryBlocks stores memory assignments of a container
+	SetMemoryBlocks(podUID string, containerName string, blocks []Block)
+	// SetMemoryAssignments sets ContainerMemoryAssignments by using the passed parameter
+	SetMemoryAssignments(assignments ContainerMemoryAssignments)
+	// Delete deletes corresponding Blocks from ContainerMemoryAssignments
+	Delete(podUID string, containerName string)
+	// ClearState clears machineState and ContainerMemoryAssignments
+	ClearState()
+}
+
+// State interface provides methods for tracking and setting memory/pod assignment
+type State interface {
+	Reader
+	writer
+}
+
+### Changes of existing components
+
+#### Container Manager changes
+
+Container Manager will create Memory Manager and register to Topology Manager as hint provider.
+
+#### Topology Manager changes
+
+Topology Manager will call out Memory Manager to gather topology hint and allocate memory resources during its admit sequence.
+
+#### Internal Container Lifecycle changes
+
+InternalContainerLifecycle will call out Memory Manager on container lifecycle events(Add/Stop) to allocate and reclaim memory resources.
+
+### Feature Gate and Kubelet Flags
+
+A new feature gate will be added to enable the Memory Manager feature. This feature gate will be disabled by default in the early releases.
+
+#### Feature Gate  
+`--feature-gate=MemoryManager=true`
+
+#### Memory Manager Policy Flag  
+`--memory-manager-policy` - to configure Memory Manager behaviour, `static` enables multi-NUMA (and single-NUMA) feature, whereas `none` disables it 
+
+Example:
+`--memory-manager-policy=none` or `--memory-manager-policy=static`
+
+#### Pre-Reserved Memory Zone Flag
+
+This will be also followed by a Kubelet Flag for the Memory Manager configuration, which is described above in [The configuration of Memory Manager](#the-configuration-of-memory-manager) section.
+
+`--pre-reserved-memory-zone=[{numa-node=int, memory-type=string, limit=string}][,][...]`
+
+Example:
+   `--pre-reserved-memory-zone= {numa-node=0, memory-type=memory, limit=1Gi}, {numa-node=1, memory-type=memory, limit=1Gi}`
+
+#### Multi-NUMA Groups Flag
+`--multi-numa-groups` - to configure disjoint groups of NUMA nodes, used for preferred multi-NUMA node memory allocation
+
+Example:
+`--multi-numa-groups=[[0, 4], [1, 5]]`
+
+<!--
+By design, *Multi-NUMA* strategy must treat a group (a group of NUMA nodes) as a unit. *Multi-NUMA* strategy will allocate the memory from more than one NUMA node if needed. 
+-->
+
+The groups must be disjoint, however, as shown in the figure in [this](#design-overview) section. For instance, `[0]`, `[1, 2]`, `[3]` groups are disjoint. This is because the groups that overlap do not ensure guaranteed memory allocation across multiple NUMA nodes (e.g. `[0, 1]`, `[1, 2]`, `[3]`). In other words, if a pair of groups overlap, memory resources of some group can be prior consumed by another group so that the other group will starve.
```

Comment:
It should remain because this concept (disjoint groups of NUMA nodes) and `--multi-numa` are independent concepts.
Disjoint groups of NUMA nodes is something that must exist to offer guaranteed memory allocation, and  `--multi-numa` flag was something additional developed on top of that.